File: new.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

// Movie represents a movie with its title and cast.
public class Movie
{
    public string Title { get; set; }
    public List<string> Cast { get; set; }
}

// Graph represents a graph using an adjacency list.
public class Graph
{
    public Dictionary<string, List<string>> AdjacencyList { get; private set; }

    public Graph()
    {
        AdjacencyList = new Dictionary<string, List<string>>();
    }

    // AddVertex adds a vertex to the graph.
    public void AddVertex(string vertex)
    {
        if (!AdjacencyList.ContainsKey(vertex))
        {
            AdjacencyList[vertex] = new List<string>();
        }
    }

    // AddEdge adds an edge between two vertices.
    public void AddEdge(string vertex1, string vertex2)
    {
        if (!AdjacencyList.ContainsKey(vertex1)) AddVertex(vertex1);
        if (!AdjacencyList.ContainsKey(vertex2)) AddVertex(vertex2);
        AdjacencyList[vertex1].Add(vertex2);
        AdjacencyList[vertex2].Add(vertex1);
    }

    // Size returns the number of vertices in the graph.
    public int Size()
    {
        return AdjacencyList.Count;
    }

    // FindShortestPaths finds the shortest paths between two actors.
    public void FindShortestPaths(string actor1, string actor2)
    {
        var queue = new Queue<List<string>>();
        queue.Enqueue(new List<string> { actor1 });
        var visited = new HashSet<string> { actor1 };
        var paths = new List<List<string>>();

        while (queue.Count > 0)
        {
            var currentPath = queue.Dequeue();
            var currentVertex = currentPath[^1];

            // Limit path length to 8 edges (9 vertices)
            if (currentPath.Count > 9)
                continue;

            if (currentVertex == actor2)
            {
                paths.Add(new List<string>(currentPath));
                Console.WriteLine(paths.Count);
                continue;
            }

            foreach (var neighbor in AdjacencyList[currentVertex])
            {
                if (!visited.Contains(neighbor) || currentPath.Count < 9)
                {
                    var newPath = new List<string>(currentPath) { neighbor };
                    queue.Enqueue(newPath);

                    if (newPath.Count < 9)
                    {
                        visited.Add(neighbor);
                    }
                }
            }
        }

        if (paths.Count > 0)
        {
            Console.WriteLine($"Shortest paths between {actor1} and {actor2}:");
            Console.WriteLine(paths.Count);
            // foreach (var path in paths)
            // {
            //     Console.WriteLine(string.Join(" -> ", path));
            // }
        }
        else
        {
            Console.WriteLine($"No relationship between {actor1} and {actor2}");
        }
    }
}

class Program
{
    static void Main()
    {
        var jsonData = File.ReadAllText("latest_movies.json");

        List<Movie> movies;
        try
        {
            movies = JsonConvert.DeserializeObject<List<Movie>>(jsonData);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error parsing JSON: {ex.Message}");
            return;
        }

        var graph = new Graph();

        foreach (var movie in movies)
        {
            graph.AddVertex(movie.Title);
            foreach (var actor in movie.Cast)
            {
                graph.AddVertex(actor);
                graph.AddEdge(movie.Title, actor);
            }
        }

        string actor1 = "Zendaya";
        string actor2 = "Shia LaBeouf";
        Console.WriteLine($"Graph size: {graph.Size()}");
        graph.FindShortestPaths(actor1, actor2);
    }
}

File: new.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
)

// Movie represents a movie with its title and cast.
type Movie struct {
	Title string   `json:"title"`
	Cast  []string `json:"cast"`
}

// Graph represents a graph using an adjacency list.
type Graph struct {
	AdjacencyList map[string][]string
}

// NewGraph initializes a new graph.
func NewGraph() *Graph {
	return &Graph{AdjacencyList: make(map[string][]string)}
}

// AddVertex adds a vertex to the graph.
func (g *Graph) AddVertex(vertex string) {
	if _, exists := g.AdjacencyList[vertex]; !exists {
		g.AdjacencyList[vertex] = []string{}
	}
}

// AddEdge adds an edge between two vertices.
func (g *Graph) AddEdge(vertex1, vertex2 string) {
	g.AdjacencyList[vertex1] = append(g.AdjacencyList[vertex1], vertex2)
	g.AdjacencyList[vertex2] = append(g.AdjacencyList[vertex2], vertex1)
}

// Size returns the number of vertices in the graph.
func (g *Graph) Size() int {
	return len(g.AdjacencyList)
}

// FindShortestPaths finds the shortest paths between two actors.
func (g *Graph) FindShortestPaths(actor1, actor2 string) {
	queue := [][]string{{actor1}}
	visited := map[string]bool{actor1: true}
	var paths [][]string

	for len(queue) > 0 {
		currentPath := queue[0]
		queue = queue[1:]
		currentVertex := currentPath[len(currentPath)-1]

		// Limit path length to 8 edges (9 vertices)
		if len(currentPath) > 9 {
			continue
		}

		if currentVertex == actor2 {
			paths = append(paths, currentPath)
			fmt.Println(len(paths))
			continue
		}

		for _, neighbor := range g.AdjacencyList[currentVertex] {
			if !visited[neighbor] || len(currentPath) < 9 {
				newPath := append([]string{}, currentPath...) // Copy currentPath
				newPath = append(newPath, neighbor)
				queue = append(queue, newPath)

				if len(newPath) < 9 {
					visited[neighbor] = true
				}
			}
		}
	}

	if len(paths) > 0 {
		fmt.Printf("Shortest paths between %s and %s:\n", actor1, actor2)
		fmt.Println(len(paths))
		// for _, path := range paths {
		// 	fmt.Println(path)
		// }
	} else {
		fmt.Printf("No relationship between %s and %s\n", actor1, actor2)
	}
}

func main() {
	data, err := ioutil.ReadFile("latest_movies.json")
	if err != nil {
		log.Fatalf("Error reading the file: %v", err)
	}

	var movies []Movie
	if err := json.Unmarshal(data, &movies); err != nil {
		log.Fatalf("Error parsing JSON: %v", err)
	}

	graph := NewGraph()

	for _, movie := range movies {
		graph.AddVertex(movie.Title)
		for _, actor := range movie.Cast {
			graph.AddVertex(actor)
			graph.AddEdge(movie.Title, actor)
		}
	}

	actor1 := "Zendaya"
	actor2 := "Shia LaBeouf"
	fmt.Printf("Graph size: %d\n", graph.Size())
	graph.FindShortestPaths(actor1, actor2)
}
